### DISCUSSION TITLE
❓ [QUESTION] Rotation equivariant layer for 3D voxel data

### DISCUSSION THREAD

**Discussion by fpschill** [e3nn/e3nn#289] [Q&A]
Hi

Thanks a lot for making this framework available!

My use case is to construct NN layers which operate on 3D voxel data (CT scans) and are equivariant with respect to rotations around a given set of axis (e.g. only around the z-axis). After studying the provided examples and documentation, it is not entirely clear if and how this is supported through e3nn (voxel data rather than point cloud, specific configuration of a z-rotation equivariant layer operating on 3D data).

Any hints would be much appreciated.

Thanks in advance, best wishes
Frank

**Comment by mariogeiger** [comment]
Hi Frank,

We really focus on O(3), rotation about any axis...
[Here](https://github.com/e3nn/e3nn/blob/main/e3nn/nn/models/v2103/voxel_convolution.py) is a file that implement a convolution for voxel.
I wrote it specifically for MRI data because I know that x-y resolution can be different than z resolution, so I added a `steps` argument that allow to have a non equiaxed grid.
Another obvious method would be to interpolate on an equiaxed grid.

...But if you really just want to be equivariant to z-axis rotation you should look at architecture that are O(2) equivariant

**Comment by blondegeek** [comment]
Hi Frank,

An easy way to get O(2) equivariance from an O(3) network is to put a
vector (or maybe it’s a pseudo vector) on each voxel pointing perpendicular
to the desired rotation plane. I would recommend that over using a strictly
O(2) network.

Like Mario said, you can use e3nn for voxels — we just haven’t focused on
it first because there are more technicalities with unstructured geometry
(graphs, point clouds). If you do start using it there’s a good chance we
can help optimize it together.

Let us know if you have any other questions.

Best,
Tess

**Comment by fpschill** [comment]
Thanks both for your quick  and helpful replies! - I'll look at the code Mario provided and try to apply it to my use-case.

Best
Frank

**Comment by blondegeek** [comment]
Hi Frank,

Depending on whether you want symmetry along the scan direction (does the
direction along the z axis matter or not) you may actually want an L=2
feature to distinguish the O(2) plane. Basically you can use L=2 features
to indicate double header rays whereas L=1 features have a definite
direction.

Best!
Tess
